API Console endpoint catalog still keys the Storage group by the deprecated `file-storage` slot — move to the canonical `storage` key (framework #9683)

Found while implementing objectstack-ai/objectstack#9683 (maintainer ruling 2026-08-18: the canonical `CoreServiceName` slot is now `storage`; `file-storage` stays a deprecated alias within v17 and retires at the next major).

## What is affected here

`apps/console/src/pages/developer/hooks/useApiDiscovery.ts` keys `SERVICE_ENDPOINT_CATALOG` by `'file-storage'` and reads `discovery.services['file-storage']` (the shape objectui#4240 fixed, correctly, against the pre-rename discovery document). `useApiDiscovery.test.ts` pins that key throughout.

## Why it keeps working today — and until when

The framework PR keeps both discovery producers mirroring the canonical `services.storage` row **verbatim** under the `file-storage` key for the alias's v17 lifetime, precisely so this console keeps rendering the Storage group unchanged. Nothing breaks now.

The mirror row goes away when the alias retires at the next major, at which point the current key would reproduce objectui#4240 exactly (fail-closed branch hides the Storage group on every host).

## Suggested resolution

Read `services.storage` first and fall back to `services['file-storage']` (so a dev console still works against an older backend), and update the test pins. The `SLOTS.has(...)` assertion can pin both spellings while the alias exists.

Not addressed in the framework PR — the console fix lands in this repository. #4240 remains closed; this is the rename follow-up, not a regression of it.

Refs: objectstack-ai/objectstack#9683, objectui#4240

---
_Generated by [Claude Code](https://claude.ai/code/session_016D9wdJR14KKCxz1WgdAzcw)_

<!-- os-dev-report -->
```json
{
  "issue": 5286,
  "status": "done",
  "branch": "claude/issue-5286-storage-canonical-slot-key",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5290",
  "premise_still_valid": true,
  "summary": "useApiDiscovery.ts's SERVICE_ENDPOINT_CATALOG lookup now reads the canonical services.storage discovery key first and falls back to the deprecated services['file-storage'] key, so the Storage group renders from either spelling while the framework's v17 alias lives (framework#9683). SERVICE_ENDPOINT_CATALOG itself stays keyed by file-storage: verified against the installed @objectstack/spec@17.0.0 that CoreServiceName.options does not export storage yet, so re-keying now would fail the existing #4240 tripwire against real types; a new exported DEPRECATED_SERVICE_SLOT_ALIASES table bridges the gap instead. Both spellings are pinned in useApiDiscovery.test.ts, reverse-verified, and a real patch changeset was added for @object-ui/console.",
  "tests": "All run from repo root per AGENTS.md's vitest-invocation guard, HEAD 47cc35bfc. `pnpm exec vitest run apps/console/src/pages/developer/hooks/useApiDiscovery.test.ts` -> 14 passed (14). Reverse verification: committed the fix, then `git checkout 12841b617 -- .../useApiDiscovery.ts` (pre-fix, kept new tests), predicted 3 of 14 would fail; observed exactly 3 failed / 11 passed -- 'renders the Storage group from the canonical `storage` slot', 'reads the canonical `storage` slot before the deprecated one when both are present' (both AssertionError: expected undefined to be defined), and 'maps the deprecated `file-storage` catalog key to the canonical `storage` discovery key' (TypeError: Cannot read properties of undefined, since DEPRECATED_SERVICE_SLOT_ALIASES doesn't exist pre-fix). Restored the fix via `git checkout claude/issue-5286-storage-canonical-slot-key -- .../useApiDiscovery.ts`; git status confirmed a clean tree matching HEAD; re-ran clean (14 passed). Also: `pnpm --filter '@object-ui/console' type-check` exit 0; `pnpm --filter '@object-ui/console' lint` 0 errors (200 pre-existing warnings, none on new lines); `node scripts/check-control-bytes.mjs` OK; `node scripts/check-changeset-presence.mjs` confirms .changeset/storage-canonical-slot-key-console.md covers both changed source files.",
  "open_questions": [],
  "out_of_scope_findings": ["filed as #5291: re-key the API Console storage catalog to `storage` and delete the DEPRECATED_SERVICE_SLOT_ALIASES bridge once the locally pinned @objectstack/spec dependency publishes `storage` as a CoreServiceName member -- dependency-bump cleanup, not a live defect"]
}
```


---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/storage-canonical-slot-key-console.md
+++ b/.changeset/storage-canonical-slot-key-console.md
@@ -1,0 +1,18 @@
+---
+'@object-ui/console': patch
+---
+
+API Console's endpoint catalog reads the canonical `storage` service slot
+first, falling back to the deprecated `file-storage` slot while the
+framework's v17 alias lives (objectui#5286, framework#9683).
+
+`useApiDiscovery`'s `SERVICE_ENDPOINT_CATALOG` looks its service names up
+directly in `/discovery`'s `services` map, and that lookup is deliberately
+fail-closed (ADR-0076 D12) — a miss hides the whole endpoint group rather
+than erroring. The framework's #9683 ruling made `storage` the canonical
+`CoreServiceName` slot and kept `file-storage` mirrored, byte-equal,
+alongside it for `@objectstack/spec`'s v17 lifetime. Before this change the
+console only ever read the deprecated key; it now reads the canonical key
+first, so the Storage group correctly reflects the framework's own naming,
+and still renders unchanged against a backend that has not deployed the
+#9683 mirror row yet.

--- a/apps/console/src/pages/developer/hooks/useApiDiscovery.test.ts
+++ b/apps/console/src/pages/developer/hooks/useApiDiscovery.test.ts
@@ -1,20 +1,26 @@
 // Copyright (c) 2026 ObjectStack. Licensed under the Apache-2.0 license.
 
 /**
- * useApiDiscovery — service-gated endpoint groups (#4240).
+ * useApiDiscovery — service-gated endpoint groups (#4240, #5286).
  *
- * ## The defect this file pins
+ * ## The defect #4240 pinned, and how #9683 reshaped it
  *
- * `SERVICE_ENDPOINT_CATALOG` keys are looked up straight in `/discovery`'s
- * `services` map, which the framework keys by `CoreServiceName`. The storage
- * group was keyed `storage`; the canonical slot is `file-storage`. So the
- * lookup missed on every host, and a miss is indistinguishable from "no such
- * service" — the deliberate fail-closed branch (ADR-0076 D12) then hid all
- * three storage endpoints everywhere, on every deployment, forever.
+ * `SERVICE_ENDPOINT_CATALOG` keys are looked up in `/discovery`'s `services`
+ * map, which the framework keys by `CoreServiceName`. #4240 fixed a mis-key
+ * (the storage group was keyed `storage` when the slot was `file-storage`)
+ * by moving the catalog to `file-storage` — the then-canonical spelling. A
+ * miss here is indistinguishable from "no such service" — the deliberate
+ * fail-closed branch (ADR-0076 D12) then hides the group everywhere, on
+ * every deployment, silently.
  *
- * That is the nastiest shape this page can fail in: fail-closed is *correct*
- * behaviour, so the symptom is silence rather than an error, and a mis-keyed
- * group is indistinguishable from a legitimately absent service.
+ * The framework's #9683 ruling (2026-08-18) inverted the slot's own
+ * canonical spelling: `storage` is now canonical, and `file-storage` is a
+ * deprecated alias both discovery producers mirror byte-equal for the
+ * alias's v17 lifetime. #5286 is the consumer-side follow-up pinned here:
+ * the lookup now reads `services.storage` first and falls back to
+ * `services['file-storage']`, so this page keeps rendering the group
+ * against BOTH an upgraded backend and one that predates the #9683 mirror
+ * row — never regressing to the #4240 failure on either spelling.
  *
  * ## What is asserted, and why the gate is left real
  *
@@ -28,6 +34,16 @@
  * The last block is the tripwire the fix would have needed to be caught: it
  * derives the expected key vocabulary from the spec itself rather than
  * restating it, so a rename on EITHER side goes red instead of going quiet.
+ * It still pins `SERVICE_ENDPOINT_CATALOG` to `file-storage`, not `storage`:
+ * the locally pinned `@objectstack/spec` dependency has not published
+ * `storage` as a `CoreServiceName` member yet (verified 2026-08-19 —
+ * `CoreServiceName.options` from the installed package lists `file-storage`
+ * only), so asserting `storage` membership here would fail against real,
+ * currently-installed types, not against the fix. See
+ * `DEPRECATED_SERVICE_SLOT_ALIASES`'s doc comment in the source file for the
+ * re-key-when-the-dependency-catches-up plan; the last `it` in that block
+ * pins the alias table directly instead, since it needs no spec version to
+ * be true.
  */
 
 import { describe, it, expect, vi, afterEach } from 'vitest';
@@ -37,6 +53,7 @@ import { CoreServiceName } from '@objectstack/spec/system';
 import {
   useApiDiscovery,
   SERVICE_ENDPOINT_CATALOG,
+  DEPRECATED_SERVICE_SLOT_ALIASES,
   type EndpointGroup,
 } from './useApiDiscovery';
 
@@ -83,17 +100,45 @@ afterEach(() => {
   vi.unstubAllGlobals();
 });
 
-describe('useApiDiscovery — the storage group (#4240)', () => {
-  it('renders the Storage group with its 3 endpoints when /discovery reports the `file-storage` slot usable', async () => {
-    const groups = await groupsFor({ 'file-storage': USABLE_STORAGE });
+describe('useApiDiscovery — the storage group (#4240, #5286)', () => {
+  it('renders the Storage group from the canonical `storage` slot (framework #9683)', async () => {
+    const groups = await groupsFor({ storage: USABLE_STORAGE });
 
     const storage = storageGroup(groups);
-    expect(storage, 'Storage group must render when the file-storage slot is filled and usable').toBeDefined();
+    expect(storage, 'Storage group must render when /discovery reports the canonical `storage` slot usable').toBeDefined();
     expect(storage!.endpoints.map(e => `${e.method} ${e.path}`)).toEqual([
       'POST /api/v1/storage/upload',
       'GET /api/v1/storage/:fileId',
       'DELETE /api/v1/storage/:fileId',
     ]);
+  });
+
+  // The v17 alias still resolves — a hard swap to `storage`-only would pass
+  // the case above but fail this one, and would break this console against
+  // any backend that has not deployed the #9683 mirror row yet.
+  it('still renders the Storage group from the deprecated `file-storage` slot while the v17 alias lives', async () => {
+    const groups = await groupsFor({ 'file-storage': USABLE_STORAGE });
+
+    const storage = storageGroup(groups);
+    expect(storage, 'Storage group must keep rendering against a backend that has not deployed the #9683 mirror row yet').toBeDefined();
+    expect(storage!.endpoints.map(e => `${e.method} ${e.path}`)).toEqual([
+      'POST /api/v1/storage/upload',
+      'GET /api/v1/storage/:fileId',
+      'DELETE /api/v1/storage/:fileId',
+    ]);
+  });
+
+  // In production the two rows are byte-equal (the framework mirrors one
+  // into the other), so this scenario cannot occur against a real backend —
+  // it exists only to pin READ ORDER: canonical-first, not an OR/merge of
+  // the two. A `file-storage`-first (or merged) lookup would read the
+  // disabled row here and fail-close.
+  it('reads the canonical `storage` slot before the deprecated one when both are present', async () => {
+    const groups = await groupsFor({
+      storage: USABLE_STORAGE,
+      'file-storage': { enabled: false },
+    });
+    expect(storageGroup(groups), 'canonical `storage` must be read first, not merged with `file-storage`').toBeDefined();
   });
 
   it('honours the route /discovery advertises for the slot, not just the default', async () => {
@@ -106,15 +151,6 @@ describe('useApiDiscovery — the storage group (#4240)', () => {
       '/api/v2/files/:fileId',
       '/api/v2/files/:fileId',
     ]);
-  });
-
-  // The canonical name is the ONLY accepted spelling. A tolerant reader here
-  // (accepting `storage` as well) would be a consumer-side workaround for a
-  // producer-side vocabulary, and would re-hide the real defect the moment the
-  // framework changed anything.
-  it('does not resurrect the group from the legacy `storage` spelling', async () => {
-    const groups = await groupsFor({ storage: USABLE_STORAGE });
-    expect(storageGroup(groups)).toBeUndefined();
   });
 });
 
@@ -178,5 +214,16 @@ describe('SERVICE_ENDPOINT_CATALOG keys are canonical service-slot names', () =>
         + 'and their groups will never render on any host — key them by the canonical slot name, '
         + 'or retire the entry',
     ).toEqual([]);
+  });
+
+  // #5286: the DISCOVERY LOOKUP already prefers the canonical `storage` wire
+  // key (framework #9683) even though the catalog's own key vocabulary above
+  // has not been re-keyed yet (see the comment on the `file-storage` catalog
+  // entry in the source file). Pin the alias table directly rather than
+  // through `CoreServiceName.options`, which does not yet list `storage` in
+  // the locally pinned `@objectstack/spec` dependency — this assertion does
+  // not depend on that dependency catching up.
+  it('maps the deprecated `file-storage` catalog key to the canonical `storage` discovery key', () => {
+    expect(DEPRECATED_SERVICE_SLOT_ALIASES['file-storage']).toBe('storage');
   });
 });

--- a/apps/console/src/pages/developer/hooks/useApiDiscovery.ts
+++ b/apps/console/src/pages/developer/hooks/useApiDiscovery.ts
@@ -200,13 +200,28 @@ export const SERVICE_ENDPOINT_CATALOG: Record<string, { group: string; defaultRo
   },
   // The KEY is the canonical service-slot name, because it is looked up
   // straight in `/discovery`'s `services` map (`discoveredServices[serviceName]`
-  // below) — and that map is keyed by `CoreServiceName`. The slot is
+  // below) — and that map is keyed by `CoreServiceName`. The slot WAS
   // `file-storage`; the ROUTE is `/api/v1/storage`. Keying this `storage` (as it
   // was until #4240) missed on every host, and the miss is indistinguishable
   // from "no such service" — so the deliberate fail-closed branch hid all three
   // endpoints everywhere, forever. The group's display name stays `Storage`:
   // that is the route's name and the user-facing one, and it is not derived
   // from the key.
+  //
+  // #5286: the framework's #9683 ruling (2026-08-18) inverted the slot's own
+  // canonical spelling — `storage` is now canonical and `file-storage` is a
+  // deprecated alias both discovery producers mirror byte-equal for the
+  // alias's v17 lifetime. This entry stays keyed `file-storage` rather than
+  // being re-keyed to `storage`: the locally pinned `@objectstack/spec`
+  // dependency has not published `storage` as a `CoreServiceName` member yet
+  // (verified 2026-08-19 against the installed package — see
+  // `DEPRECATED_SERVICE_SLOT_ALIASES` below), so re-keying now would fail the
+  // "every service-gated catalog key is a canonical slot" tripwire in the
+  // test file against the real, currently-installed types. What DOES change
+  // here is the discovery lookup: it reads the canonical `services.storage`
+  // first and falls back to this key, so the group renders from either
+  // spelling. Re-key this entry to `storage` and delete the alias table
+  // once the dependency catches up.
   'file-storage': {
     group: 'Storage',
     defaultRoute: '/api/v1/storage',
@@ -216,6 +231,26 @@ export const SERVICE_ENDPOINT_CATALOG: Record<string, { group: string; defaultRo
       { method: 'DELETE', path: '/:fileId', desc: 'Delete file' },
     ],
   },
+};
+
+/**
+ * Deprecated → canonical `CoreServiceName` slot aliases (framework #9683,
+ * maintainer ruling 2026-08-18: `storage` is now the canonical slot; the
+ * compound `file-storage` spelling is a deprecated alias kept for
+ * `@objectstack/spec`'s v17 lifetime). `/discovery`'s `services` map reports
+ * BOTH keys — byte-equal — for as long as the alias lives, so
+ * {@link useApiDiscovery}'s lookup below reads the canonical key first and
+ * falls back to the deprecated one only when the canonical key is entirely
+ * absent (a backend that predates the #9683 mirror row). Both spellings are
+ * pinned in `useApiDiscovery.test.ts`.
+ *
+ * This table is keyed by the CATALOG's key (still `file-storage`, see the
+ * comment on that entry above), not by every `CoreServiceName` value — add
+ * an entry here only when a catalog key itself has a deprecated spelling to
+ * bridge.
+ */
+export const DEPRECATED_SERVICE_SLOT_ALIASES: Record<string, string> = {
+  'file-storage': 'storage',
 };
 
 export function buildServiceEndpoints(serviceName: string, routePrefix: string): EndpointDef[] {
@@ -270,9 +305,13 @@ export function useApiDiscovery() {
 
       const serviceEndpoints: EndpointDef[] = [];
       for (const [serviceName, catalog] of Object.entries(SERVICE_ENDPOINT_CATALOG)) {
-        const serviceInfo = discoveredServices[serviceName] as
-          | (DiscoveryServiceStatus & { route?: string })
-          | undefined;
+        // #5286: read the canonical slot name first (framework #9683) and
+        // fall back to the catalog's own (possibly deprecated) key — see
+        // `DEPRECATED_SERVICE_SLOT_ALIASES`'s doc comment above.
+        const canonicalSlot = DEPRECATED_SERVICE_SLOT_ALIASES[serviceName];
+        const serviceInfo = (
+          (canonicalSlot ? discoveredServices[canonicalSlot] : undefined) ?? discoveredServices[serviceName]
+        ) as (DiscoveryServiceStatus & { route?: string }) | undefined;
         // ADR-0076 D12 (framework#2462): gate on the shared honest-capability
         // check — `stub`/`unavailable`/`handlerReady:false` entries must not
         // render endpoint groups (`degraded` still serves and stays visible).

